### PR TITLE
Auth Adapters refactoring

### PR DESCRIPTION
**Pull request by flovilmart**
See issue #3175

Auth providers now follow the adapter pattern and can be specified as adapters with multiple possible options.

Moved all auth data to the providers
Adds tests for loading adapters from multiple sources (code, module, function module etc..)

### CODE REVIEW

File: src/ParseServer.js [hunk 1]
```diff
@@ -191,6 +189,17 @@ class ParseServer {
 
     const dbInitPromise = databaseController.performInitialization();
 
+    if (Object.keys(oauth).length > 0) {
+      /* eslint-disable no-console */
+      console.warn('oauth option is deprecated and will be removed in a future release, please use auth option instead');
```
Comment (acinader) [inline src/ParseServer.js]: is it possible or desirable to use the logger here?  Maybe you just expect users to see this during dev/testing?  reasonable i guess, but would still see with the logger.  maybe logger isn't available yet?

File: src/ParseServer.js [hunk 2]
```diff
@@ -191,6 +189,17 @@ class ParseServer {
...
+      if (Object.keys(auth).length > 0) {
+        console.warn('You should your only use the auth option.');
```
Comment (acinader) [inline src/ParseServer.js]: all your bases....

File: src/ParseServer.js (hunk 1 above)
Comment (flovilmart) [inline src/ParseServer.js]: logger is not available, and probably not configured to spit to the console

File: src/ParseServer.js (hunk 2 above)
Comment (flovilmart) [inline src/ParseServer.js]: yay!

File: src/ParseServer.js (hunk 1 above)
Comment (flovilmart) [inline src/ParseServer.js]: like we have here:

```
          /* eslint-disable no-console */
          console.error(`Unable to listen on port ${err.port}. The port is already in use.`);
          /* eslint-enable no-console */
```

Comment (acinader) [inline src/ParseServer.js]: i buy it

Comment (flovilmart) [inline src/ParseServer.js]: @acinader Still pending your approval :)

**Pull request merged**